src/server/frontend_wayland/shm.cpp

```diff
@@ -0,0 +1,336 @@
+      stride_{stride},
+      format_{format}
+{
+}
+
+namespace
+{
+class ErrorNotifyingRWMappableBuffer : public mrs::RWMappableBuffer
+{
+public:
+    ErrorNotifyingRWMappableBuffer(
+        mir::wayland::Weak<mf::ShmBuffer> buffer,
+        std::shared_ptr<mir::Executor> wayland_executor,
+        std::shared_ptr<mir::shm::RWMappableRange> data,
+        mir::geometry::Size size,
+        mir::geometry::Stride stride,
+        MirPixelFormat format);
+
+    auto size() const -> mir::geometry::Size override;
+    auto stride() const -> mir::geometry::Stride override;
+    auto format() const -> MirPixelFormat override;
+
+    auto map_readable() -> std::unique_ptr<mrs::Mapping<unsigned char const>> override;
+    auto map_writeable() -> std::unique_ptr<mrs::Mapping<unsigned char>> override;
+    auto map_rw() -> std::unique_ptr<mrs::Mapping<unsigned char>> override;
+
+    void notify_access_error() const;
+private:
+    mir::wayland::Weak<mf::ShmBuffer> const weak_buffer;
+    std::shared_ptr<mir::Executor> const wayland_executor;
+    std::shared_ptr<mir::shm::RWMappableRange> const data;
+    mir::geometry::Size const size_;
+    mir::geometry::Stride const stride_;
+    MirPixelFormat const format_;
+};
+
+template<typename T>
+class ErrorNotifyingMapping : public mir::renderer::software::Mapping<T>
+{
+public:
+    ErrorNotifyingMapping(
+        std::unique_ptr<mir::shm::Mapping<std::conditional_t<std::is_const_v<T>, std::byte const, std::byte>>> mapping,
+        ErrorNotifyingRWMappableBuffer const& parent)
+        : mapping{std::move(mapping)},
+          parent{parent}
+    {
+    }
+
+    ~ErrorNotifyingMapping()
+    {
+        if (mapping->access_fault())
+        {
+            parent.notify_access_error();
+        }
+    }
+
+    auto format() const -> MirPixelFormat override
+    {
+        return parent.format();
+    }
+
+    auto stride() const -> mir::geometry::Stride override
+    {
+        return parent.stride();
+    }
+
+    auto size() const -> mir::geometry::Size override
+    {
+        return parent.size();
+    }
+
+    auto data() -> T* override
+    {
+        return reinterpret_cast<T*>(mapping->data());
+    }
+
+    auto len() const -> size_t override
+    {
+        return mapping->len();
+    }
+
+private:
+    std::unique_ptr<mir::shm::Mapping<std::conditional_t<std::is_const_v<T>, std::byte const, std::byte>>> const mapping;
+    ErrorNotifyingRWMappableBuffer const& parent;
+};
+
+ErrorNotifyingRWMappableBuffer::ErrorNotifyingRWMappableBuffer(
+        mir::wayland::Weak<mf::ShmBuffer> buffer,
+        std::shared_ptr<mir::Executor> wayland_executor,
+        std::shared_ptr<mir::shm::RWMappableRange> data,
+        mir::geometry::Size size,
+        mir::geometry::Stride stride,
+        MirPixelFormat format)
+    : weak_buffer{buffer},
+      wayland_executor{std::move(wayland_executor)},
+      data{std::move(data)},
+      size_{std::move(size)},
+      stride_{stride},
+      format_{format}
```

Comment:
Not our usual indentation

Reply:
@AlanGriffiths I've fixed, is that what you had in mind? 7631fb1